packages/@atjson/document/src/index.ts

```diff
@@ -80,139 +91,34 @@ export default class AtJSON {
-      // comments, where insertion at the end of the link/comment should _not_
-      // affect the annotation.
-      //
-      // FIXME this whole inner loop should probably be moved to a base Annotation.transform
-      if (a.transform) {
-        a.transform(a, this.content, position, text.length, preserveAdjacentBoundaries);
-
-      // The first two normal cases are self explanatory. Just adjust the annotation
-      // position, since there is never a case where we wouldn't want to.
-      } else if (position < a.start) {
-        a.start += length;
-        a.end += length;
-      } else if (position > a.start && position < a.end) {
-        a.end += length;
-
-      // In this case, however, the normal behaviour when inserting text at a
-      // point adjacent to an annotation is to drag along the end of the
-      // annotation, or push forward the beginning, i.e., the transform happens
-      // _inside_ an annotation to the left, or _outside_ an annotation to the right.
-      //
-      // Sometimes, the desire is to change the direction; this is provided below
-      // with the preserveAdjacentBoundaries switch.
+  insertText(start: number, text: string, behaviour: AdjacentBoundaryBehaviour = AdjacentBoundaryBehaviour.default) {
+    if (start < 0 || start > this.content.length) throw new Error('Invalid position.');
 
-      // Default edge behaviour.
-      } else if (!preserveAdjacentBoundaries) {
-        if (position === a.start) {
-          a.start += length;
-          a.end += length;
-        } else if (position === a.end) {
-          a.end += length;
-        }
-
-      // Non-standard behaviour. Do nothing to the adjacent boundary!
-      } else if (position === a.start) {
-        a.end += length;
-
-      // no-op; we would delete the annotation, but we should defer to the
-      // annotation as to whether or not it's deletable, since some zero-length
-      // annotations should be retained.
-      // n.b. the += 0 is just to silence tslint ;-)
-      } else if (position === a.end)  {
-        a.end += 0;
+    let insertion = new Insertion(start, text, behaviour);
+    try {
+      for (let i = this.annotations.length - 1; i >= 0; i--) {
+        let annotation = this.annotations[i];
+        annotation.handleChange(insertion);
       }
-    }
-  }
-
-  deleteText(annotation: Annotation) {
-    // This should really not just truncate annotations, but rather tombstone
-    // the modified annotations as an atjson sub-document inside the annotation
-    // that's being used to delete stuff.
 
-    const start = annotation.start;
-    const end = annotation.end;
-    const length = end - start;
-
-    if (!(start >= 0 && end >= 0)) {
-      throw new Error('Start and end must be numbers.');
+      this.content = this.content.slice(0, start) + text + this.content.slice(start);
+    } catch (e) {
+      // tslint:disable-next-line:no-console
+      console.error('Failed to insert text', e);
```

Comment:
Q: why console instead of throw?

Reply:
Probably should throw!

Reply:
Hmm. I saw the context for this again. I think we're doing this to guard from non-atjson code from causing apps to blow up. We will revisit this, for sure.

Reply:
Makes sense. As a possible follow-up for another PR, maybe own your own logger interface. Then you can control if/when to output and to where.